### DISCUSSION TITLE
Access messages quickly

### DISCUSSION THREAD

**Discussion by eriolloan** [andyholmes/valent#689] [Feature Request]
Is there a way to allow direct access to the messaging window ?

currently if you want to start/read a conversation it requires three steps : opening the application window > select the device > select "Messages".

What about adding a .desktop file to open the message window directly (and potentially add it to the dash for easy access)?

**Comment by andyholmes** [comment]
That's a good idea, although I'm not sure how to do it exactly.

Each device is a `GActionGroup` (with an action to open the messaging window) and gets exported on D-Bus, but you'd have to know the devices, which supported messages and which to open.

Probably it makes sense to have a shared window, like with the media remote. The window could either organize all messages or have dropdown to switch between devices.

**Reply by eriolloan** [reply]
Your ideas and remarks sparked my imagination toward a combined message list and I came up with a possible UI.
I took inspiration from the Calendar app and other existing patterns and tried to take the concerns I could think of into account. I may have overlooked some aspects.

To differentiate devices I went the color-coding route. However it would probably rest upon some UI changes in the main window.

Here's a proposition : 

![valent-standalone-message-window](https://github.com/andyholmes/valent/assets/15250783/d24cdd21-f386-4183-b7b9-ccf67fd5e359)


___________________

> Each device is a `GActionGroup` (with an action to open the messaging window) and gets exported on D-Bus, but you'd have to know the devices, which supported messages and which to open.

I'm sorry but I really hope this is only code-related and not impacting the UI as this reads like Chinese to me as my skill are limited to taking part to design sprints and producing mockups.


**edit:** replaced svg with png as it wouldn't render properly.

**Reply by andyholmes** [reply]
Thanks, that looks really great!

I think the color preference probably needs some thought, since it goes beyond the SMS plugin, but the initial design could be implemented with labels. Changes to the device page would also need more thought, since it's plugin-based and this would require a static layout.

> I'm sorry but I really hope this is only code-related and not impacting the UI as this reads like Chinese to me as my skill are limited to taking part to design sprints and producing mockups.

This is reason why there's no easy, direct access to the messaging window. That's all.

**Reply by eriolloan** [reply]
Thanks !

Can you elaborate about the labels?
As for the device page, what are the design implications of it being plugin-based, and what limitations do I need to have in mind when designing a possible UI?

**Reply by andyholmes** [reply]
> Can you elaborate about the labels?

Anywhere you're using color to distinguish devices, you'd have to use labels until something like a color preference can be thought out. This would have to happen anyways, since using color as the only distinguishing attribute is an accessibility issue.

> As for the device page, what are the design implications of it being plugin-based, and what limitations do I need to have in mind when designing a possible UI?

Everything is built programmatically, which is why the current device page is basically two lists: the list of gadgets (battery, mobile network), and the list of actions.

**Reply by eriolloan** [reply]
I found no tickets regarding colors and accessibility on GNOME Calendar's GitLab. As a Core app, it seems this has been deemed a non-issue at multiple levels (devs, users and curators). The app generates colors but doesn't lock them in, allowing visually impaired users to choose contrasting colors to suit their needs.

But as you mentioned, if nothing can be done currently about the device page, the feature is not discoverable enough and using labels is probably the best option for now.

Making a mockup right now !

**Reply by eriolloan** [reply]
![valent-standalone-message-window-with-labels](https://github.com/andyholmes/valent/assets/15250783/50dfe9d2-e3df-4196-a133-bfbe9927543f)

**Reply by andyholmes** [reply]
> The app generates colors but doesn't lock them in, allowing visually impaired users to choose contrasting colors to suit their needs.

Right, this affects both color blindness and those who need screen readers. Color is fine, but color as the only distinguishing feature would be invisible to a screen reader or braille device.

Thanks this looks good!

**Reply by eriolloan** [reply]
> color as the only distinguishing feature would be invisible to a screen reader or braille device.

I'm briefly going off-topic but since I have en interest in accessibility and you mention it, does Gtk4 provide some equivalent to HTML aria-labels?

> Thanks this looks good!

I'm glad you like it! 
Looking at it now I think the message list might gain in readability if the contact names were bold... let me know if there's anything else you need or any adjustments you'd like me to make.

**Reply by andyholmes** [reply]
> I'm briefly going off-topic but since I have en interest in accessibility and you mention it, does Gtk4 provide some equivalent to HTML aria-labels?

There are equivalents, label and description, that can be applied to widgets. This would need a custom widget that's picked up by the AT device

> Looking at it now I think the message list might gain in readability if the contact names were bold... let me know if there's anything else you need or any adjustments you'd like me to make.

There might be prior art for messaging in the GNOME design team's mockup repo, although they may not apply to this narrow use case.

In general, the changes should really be separate, i.e. a color preference for devices is pretty orthogonal to the SMS window. I also think the ability to select more than one device at once will require a lot more thought, I'm not sure how complicated that will make things down the road.

**Reply by eriolloan** [reply]
> There are equivalents, label and description, that can be applied to widgets. This would need a custom widget that's picked up by the AT device

I don't know precisely what you're working with here but would an AT description on the contact label (or the list item) like "[contact] from [device]" be an option?

> In general, the changes should really be separate, i.e. a color preference for devices is pretty orthogonal to the SMS window.

Yes, I may have provided an ambitious design.

I get your point and I'm not trying to be pedantic here. I get the developer perspective (especially if they've got a lot on their plate ;)). I admit I'm biased toward an UX standpoint, and that's of course a different tune. The features that add the most value to the user often require transversal design to achieve consistency and ease of use. This obviously means more work given to the dev. I aimed to the best way to tackle the issue with what I know.

I guess the easy thing to do concerning the UI is to give the user the ability to select which is the current device from the dropdown to display messages from this source. That would achieve the feature (accessing messages from any device without opening the main window).

**Reply by andyholmes** [reply]
> I don't know precisely what you're working with here but would an AT description on the contact label (or the list item) like "[contact] from [device]" be an option?

Typically I think the row is the only focusable element, and you would set a relation between the row and the title/subtitle labels as the a11y label/description properties. It's possible to build a string for those properties, but I think we can avoid that just by restricting the window to one device's messages at once.

Currently the SMS plugin is basically not written. There's really just some basic scaffolding holding together what works now. So one way or the other, this will have to be done in stages. 

First the messaging functionality and local cache has to be finished, then the SMS plugin will have to be refactored to support a model where one window can access the messages for multiple devices. Presumably, the source of contacts will just follow the active device, because if that's also configurable the implementation becomes more complex as well.

So the mockup looks good, it's just that I'm really quite far away from putting a rubber-stamp on it, because of how much work is left to do.

**Reply by andyholmes** [reply]
> Presumably, the source of contacts will just follow the active device, because if that's also configurable the implementation becomes more complex as well.

This could actually use a mockup, if you're interested in the general messaging/contact UI. This is a factor I've never thought through or addressed (even in GSConnect).

**Reply by eriolloan** [reply]
Thank you for explaining the current state of the project and the necessary steps ahead.

>  if you're interested in the general messaging/contact UI.

I am. I'm not sure what you need exactly though.

> the source of contacts will just follow the active device

I only have one device so I have no experience about that. Do I understand correctly that there's only one active device at a time ?

**Reply by andyholmes** [reply]
> I am. I'm not sure what you need exactly though.

While I'd like the keep the preferences minimal, I think there are some things that are unavoidable. There's a stub menu-entry for contacts, basically because I'm not sure how to best present that.

> I only have one device so I have no experience about that. Do I understand correctly that there's only one active device at a time ?

Oh you can have many active devices, I meant active as in the "currently selected" device in the window. It maybe be possible to merge SMS/MMS for multiple devices in the UI, but I think it's wise to get them working one at a time first.

Right now we don't have an addressbook selection, so if the device doesn't share contacts you just get fallback avatars and raw phone numbers. But it is already written so that any contact source (e.g. another connected device, Evolution Data Server, GNOME Online Accounts, etc) can be used.

So say you have a Google Pixel and a OnePlus6 or something. When you switch from the Pixel to the OnePlus6 with the combobox, should the contact source change to that device's addressbook if available? Should their be a global preference for addressbook, per-device, or some attempt to merge contacts?

This whole area I have not thought out at all, but I'd like to think there's a nice UI/UX that's not a huge matrix of configurations to support :)

**Reply by eriolloan** [reply]
Sorry for the delay.

> should the contact source change to that device's addressbook if available? Should their be a global preference for addressbook, per-device, or some attempt to merge contacts?

Given your explanation, selecting multiple devices in the combobox would require a massive amount of work from you, only to potentially provide bad UX. I guess a solution to duplicate/conflicting contacts might eventually come up from GOA or something...

But in the meantime I think dealing with combined addressbooks can be avoided if we restrict the discussion list to contacts from the selected device.

dealing with messages from other active devices can be achieved by notifying the user with a toast at the bottom of the list. ("n unread messages on [devicename]").

Clicking on shell notifications (upon recieving a new message), would simply display the conversation window and select the appropriate device.

As for phones not providing the contact info, it could fallback to the ones in the contact app or another active device  if the phone numbers match (and maybe from known devices in the long-term as I presume this means you'd have implement a way to keep its contacts stored somewhere).

**Reply by andyholmes** [reply]
> I guess a solution to duplicate/conflicting contacts might eventually come up from GOA or something...

I believe we can handle this with TinySPARQL and ontologies (previously known as the Tracker framework), but it's a step-by-step thing for sure. Probably I'll test this by implementing the SMS/MMS cache with TinySPARQL, then if all goes well, refactor the contacts support so it can be used in tandem.

> But in the meantime I think dealing with combined addressbooks can be avoided if we restrict the discussion list to contacts from the selected device.
> dealing with messages from other active devices can be achieved by notifying the user with a toast at the bottom of the list. ("n unread messages on [devicename]").

:+1: & :+1:. The second part, I want to double-check that toasts are announced for AT devices, but that sounds like good compromise for now.

> Clicking on shell notifications (upon recieving a new message), would simply display the conversation window and select the appropriate device.

This might be hard, depending. The SMS & notification plugins are entirely separate, and the information you get from the notification is what you'd expect: title, body, icon, etc. So nothing definitive to like "this is a name/phone number/message in thread #", or even "this is an SMS message and not a WhatsApp notification".

An ugly fix is to just open the window if the body of a notification is clicked and we can guess it's an SMS message. Probably we can guess what device to select for the window. Longer term, the SMS plugin could take responsibility for showing notifications, but then the other notifications would have to be filtered with some heuristics.

> As for phones not providing the contact info, it could fallback to the ones in the contact app or another active device if the phone numbers match (and maybe from known devices in the long-term as I presume this means you'd have implement a way to keep its contacts stored somewhere).

:+1: Should also be do-able with TinySPARQL.

**Reply by andyholmes** [reply]
I threw up some changes that make some big moves towards the mock-up.

I'm leaning into TinySPARQL, which worked really well and let me avoid a lot of touch problems. It's pretty rough at the moment, but highlights are:

* All UI is now in the `gnome` plugin (i.e. the device page doesn't need to be dynamic anymore)
* There's a global manager for messaging sources (`ValentMessages`), like other components
* The messaging window is global with a dropdown for selecting device (see primary menu)

It's all very rough right now, but there's a little more interactive stuff to poke at now at least :)

**Reply by eriolloan** [reply]
Wow, you're going full speed on this ! I didn't think this was such a priority.
I'm here if you need someone to give feedback on a build at some point ;)

**Reply by andyholmes** [reply]
Well, it sort of looks like it :)

Switching to TinySPARQL is obviously something I should've done 2 years ago, because the plugin is pretty much finished (in high-level terms).

Something you can probably help with is all the extra UI bits that are missing. For example, you can attach and send files with messages, there's just no UI feedback for what you have attached or what size they are. If you poke around the UI you'll see where things should be and most of the features behind them are implemented.

I'd like keep it minimalist, but my priority is UX for a11y users (e.g. keyboard focus chain). It's obvious a desktop-to-phone SMS app is super cool and convenient for most people; I expect it could be a game changer for folks with e.g. mobility issues, so they're sort of the MVPs of this part of Valent.

**Reply by eriolloan** [reply]
Wow I didn't notice the update landed on my computer and it's awesome. And the conversation list uses the proper contact names.
I'll borrow a friend's phone to see how it all behaves with two devices and unknown contacts.

There are so many ways to deal with attachments...  Right now I notice they're focusable in the message view but have no action associated (pressing space or enter) is that meant to change in the end ?

I'll take a look at what's done on similar applications as I can't think of more UI bits needing attention right now.

I'm currently focused on paid projects, but I'll get to it soon.

**Reply by andyholmes** [reply]
> There are so many ways to deal with attachments... Right now I notice they're focusable in the message view but have no action associated (pressing space or enter) is that meant to change in the end ?

Oh yes, for those situations I'm planning on using the scaling-preview widget from Fractal (you can see it in their conversation list). I'm not sure about outgoing attachments; I think most messengers actually just send as soon as you attach, but I'll have to check.

> I'm currently focused on paid projects, but I'll get to it soon.

No worries! I'll probably be doing mini-sprints as time is available, so back-and-forth iteration works well for me :)
